accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,492 @@
+
+This change would resolve the request for [dotnet-publish with --use-current-rid or --infer-rid option](https://github.com/dotnet/sdk/issues/10449) without needing any additional syntax. This request aligns with the challenge we have with Dockerfiles.
+
+Portable framework-dependent apps are the default scenario today. As already stated, it's not a good default choice because it produces larger apps with niche benefits.
+
+### Enable gesture to build portable apps
+
+If rid-specific apps become the default, then we need a new gesture for portable apps. There are two (non-mutually-exclusive) models for that:
+
+- Use the existing `any` RID to mean portable.
+- Expose a new CLI arg and MSBuild property for portable apps, like  `--portable` and `<Portable>true</Portable>`.
+
+Recommendation: Start with using just the RID. Wait for feedback to enable more polished experiences to validate that portable is an important scenario.
+
+There is also an existing `win` RID. It could be used in a similar way as `any` but limited to just Windows family RIDs. ASP.NET Core developers deploying and/or developing apps on some combination of Windows x86, x64, and Arm64 may find that scenario useful. We should not proactively develop that scenario, but wait for feedback.
+
+In addition, the SDK would no longer generate apphost EXEs for portable apps by default. There already is an opt-in model for EXEs (`<UseAppHost>true</UseAppHost>`). Portable app users can rely on that.
+
+The rationale for making apphost EXEs opt-in is that they are a non-portable asset, creating a strange asymmetry in an otherwise portable app. Perhaps, we should make it easy for developers to generate an apphost only for debug builds to aid the development experience. That said, if developers want a better development experience, they could just as easily build RID-specific apps (the proposed default) for development and then build a portable app with no EXE for production where an EXE is presumably not needed.
+
+Note: Portable apps are not a meaningful option for client apps since almost all client apps require an executable launcher, and there is only ever one generated. Portable apps are only meaningful for commandline apps.
+
+Note: In theory, we could consider an option to create multiple apphosts for different RIDs, either of different names or in different directories. Again, this capability would need to be opt-in.
+
+### Enable RID folders for RID-specific and portable apps
+
+There have been some issues filed (see appendix) that suggest that managed libraries are not resilient to RID-specific libraries changing location as a result of changing build type (portable to RID-specific). We should consider adding an option that creates `runtimes` folder (for compatible RIDs) for RID-specific builds. Perhaps this challenge just needs to be investigated more.
```

Comment:
This also means that the library is not compatible with single-file. Single-file does not preserve the deep directory structure either.

We have seen both native libraries not compatible with the deep directory structure, and native libraries not compatible with the flat directory structure. We consider these issues to be a design problem with the native library.

For example: https://github.com/dotnet/runtime/issues/25670

Reply:
That resonates with me.

This is probably better docs and tools.

Reply:
Updated the text based on your feedback. https://github.com/dotnet/designs/pull/233/commits/92af8d18e7f2759bb4a1a94681de30405ae97c1e